public class junit.framework.TestFailure extends java.lang.Object
{
    protected junit.framework.Test fFailedTest;
    protected java.lang.Throwable fThrownException;

    public void <init>(junit.framework.Test, java.lang.Throwable)
    {
        junit.framework.TestFailure r0;
        junit.framework.Test r1;
        java.lang.Throwable r2;

        r0 := @this: junit.framework.TestFailure;

        r1 := @parameter0: junit.framework.Test;

        r2 := @parameter1: java.lang.Throwable;

        specialinvoke r0.<java.lang.Object: void <init>()>();

        r0.<junit.framework.TestFailure: junit.framework.Test fFailedTest> = r1;

        r0.<junit.framework.TestFailure: java.lang.Throwable fThrownException> = r2;

        return;
    }

    public junit.framework.Test failedTest()
    {
        junit.framework.TestFailure r0;
        junit.framework.Test $r1;

        r0 := @this: junit.framework.TestFailure;

        $r1 = r0.<junit.framework.TestFailure: junit.framework.Test fFailedTest>;

        return $r1;
    }

    public java.lang.Throwable thrownException()
    {
        junit.framework.TestFailure r0;
        java.lang.Throwable $r1;

        r0 := @this: junit.framework.TestFailure;

        $r1 = r0.<junit.framework.TestFailure: java.lang.Throwable fThrownException>;

        return $r1;
    }

    public java.lang.String toString()
    {
        junit.framework.TestFailure r0;
        java.lang.StringBuilder $r1, $r3, $r4, $r7;
        junit.framework.Test $r2;
        java.lang.Throwable $r5;
        java.lang.String $r6, $r8;

        r0 := @this: junit.framework.TestFailure;

        $r1 = new java.lang.StringBuilder;

        specialinvoke $r1.<java.lang.StringBuilder: void <init>()>();

        $r2 = r0.<junit.framework.TestFailure: junit.framework.Test fFailedTest>;

        $r3 = virtualinvoke $r1.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.Object)>($r2);

        $r4 = virtualinvoke $r3.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.String)>(": ");

        $r5 = r0.<junit.framework.TestFailure: java.lang.Throwable fThrownException>;

        $r6 = virtualinvoke $r5.<java.lang.Throwable: java.lang.String getMessage()>();

        $r7 = virtualinvoke $r4.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.String)>($r6);

        $r8 = virtualinvoke $r7.<java.lang.StringBuilder: java.lang.String toString()>();

        return $r8;
    }

    public java.lang.String trace()
    {
        junit.framework.TestFailure r0;
        java.io.StringWriter $r3;
        java.io.PrintWriter $r4;
        java.lang.Throwable $r5;
        java.lang.String $r6;

        r0 := @this: junit.framework.TestFailure;

        $r3 = new java.io.StringWriter;

        specialinvoke $r3.<java.io.StringWriter: void <init>()>();

        $r4 = new java.io.PrintWriter;

        specialinvoke $r4.<java.io.PrintWriter: void <init>(java.io.Writer)>($r3);

        $r5 = virtualinvoke r0.<junit.framework.TestFailure: java.lang.Throwable thrownException()>();

        virtualinvoke $r5.<java.lang.Throwable: void printStackTrace(java.io.PrintWriter)>($r4);

        $r6 = virtualinvoke $r3.<java.io.StringWriter: java.lang.String toString()>();

        return $r6;
    }

    public java.lang.String exceptionMessage()
    {
        junit.framework.TestFailure r0;
        java.lang.Throwable $r1;
        java.lang.String $r2;

        r0 := @this: junit.framework.TestFailure;

        $r1 = virtualinvoke r0.<junit.framework.TestFailure: java.lang.Throwable thrownException()>();

        $r2 = virtualinvoke $r1.<java.lang.Throwable: java.lang.String getMessage()>();

        return $r2;
    }

    public boolean isFailure()
    {
        junit.framework.TestFailure r0;
        java.lang.Throwable $r1;
        boolean $z0;

        r0 := @this: junit.framework.TestFailure;

        $r1 = virtualinvoke r0.<junit.framework.TestFailure: java.lang.Throwable thrownException()>();

        $z0 = $r1 instanceof junit.framework.AssertionFailedError;

        return $z0;
    }
}
